File: apple/LC33SearchInRotatedSortedArray.java
class Solution {
    public int search(int[] nums, int target) {
        if (nums.length == 0) return -1;
        int left = 0;
        int right = nums.length-1;
        if (target < nums[left] && target > nums[right]) return -1;
        while (left < right) {
            int mid = (left+right)/2;
            if (nums[mid] <= nums[right]) {
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid+1;
                }else {
                    right = mid;
                }
            }else {
                if (nums[left] <= target && target <= nums[mid]) {
                    right = mid;
                }else {
                    left = mid+1;
                }
            }
        }
        if (nums[left] == target) return left;
        else return -1;
    }
}

File: Linkedin/LC33SearchInRotatedSortedArray.py
class Solution(object):
    def search(self, nums, target):
        if not nums:
            return -1
        if target > nums[-1] and target < nums[0]:
            return -1
        left = 0
        right = len(nums)-1
        while left <= right:
            if right-left <= 1:
                if target in nums[left:right+1]:
                    return left if target == nums[left] else right
                return -1
            mid = (left+right)/2
            if nums[mid] == target:
                return mid
            elif nums[mid] > nums[left]:
                if nums[left] <= target < nums[mid]:
                    right = mid-1
                else:
                    left = mid+1
            elif nums[mid] < nums[right]:
                if nums[mid] < target <= nums[right]:
                    left = mid+1
                else:
                    right = mid-1
        return -1
                
